File: 2024/12/12.rb
#!/usr/bin/env ruby

require 'matrix'
V = Vector

DIRECTIONS = [V[0, -1], V[1, 0], V[0, 1], V[-1, 0]].freeze

def fill_region(grid, start, points_to_regions, region)
  x_range, y_range = [0...grid[0].size, 0...grid.size]
  positions = [start]
  plant = grid[start[1]][start[0]]
  while positions.any?
    p = positions.shift
    next if points_to_regions[p]
    points_to_regions[p] = region
    DIRECTIONS.each do |d|
      np = p + d
      next unless x_range.include? np[0] and y_range.include? np[1]
      positions << np if grid[np[1]][np[0]] == plant
    end
  end
end

def find_regions(grid)
  x_range, y_range = [0...grid[0].size, 0...grid.size]
  points_to_regions = {}
  region = 1
  x_range.each do |x|
    y_range.each do |y|
      v = V[x,y]
      next if points_to_regions[v]
      fill_region(grid, v, points_to_regions, region)
      region = region.succ
    end
  end
  regions_to_points = Hash.new { |h,k| h[k] = [] }
  points_to_regions.each do |p, r|
    regions_to_points[r] << p
  end
  regions_to_points
end

def fetch_perimeter_and_area(regions_to_points)
  regions_to_points.transform_values do |ps|
    ps_set = ps.to_set
    perimeter = 0
    ps.each do |p|
      DIRECTIONS.each do |d|
        perimeter = perimeter.succ unless ps_set.include?(p + d)
      end
    end
    [perimeter, ps.size]
  end
end

def fetch_sides_and_area(regions_to_points)
  regions_to_points.transform_values do |ps|
    # Walk all edges until you reach every corner
    [ps.inject(Set.new) do |corners, p|
      DIRECTIONS.each do |d|
        next if ps.include?(p + d) # not an edge side
        step = p
        dv = V[d[1], -d[0]] # rotate counter-clockwise
        # walk until you reach a corner
        step += dv while ps.include?(step + dv) && !ps.include?(step + d + dv)
        corners << [step, d]
      end
      corners
    end.size, ps.size]
  end
end

grid = []
$stdin.readlines.each_with_index do |line|
  grid << line.chomp.chars
end

regions_to_points = find_regions(grid)
part1 = fetch_perimeter_and_area(regions_to_points).inject(0) do |sum, (_r, (p, a))|
  sum + p * a
end
puts part1

part2 = fetch_sides_and_area(regions_to_points).inject(0) do |sum, (_r, (s, a))|
  sum + s * a
end
puts part2

File: 2024/12/12.go
package main

import (
	"bufio"
	"os"
)

type P struct {
	x int
	y int
}

func (this *P) add(other P) P {
	return P{this.x + other.x, this.y + other.y}
}

type C struct {
	p P
	d P
}

type Grid []Row
type Row []rune
type PointSet map[P]bool

var DIRECTIONS = []P{{0, -1}, {1, 0}, {0, 1}, {-1, 0}}

func parseInput() Grid {
	grid := Grid{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		grid = append(grid, Row(scanner.Text()))
	}
	return grid
}

func findRegions(grid Grid) []PointSet {
	pointsToRegions := map[P]int{}
	region := 1
	for y, row := range grid {
		for x := range row {
			p := P{x, y}
			if pointsToRegions[p] > 0 {
				continue
			}
			fillRegion(grid, p, pointsToRegions, region)
			region++
		}
	}
	regions := make([]PointSet, region-1)
	for i := range regions {
		regions[i] = PointSet{}
	}
	for p, r := range pointsToRegions {
		regions[r-1][p] = true
	}
	return regions
}

func fillRegion(grid Grid, start P, pointsToRegions map[P]int, region int) {
	xmax, ymax := len(grid[0])-1, len(grid)-1
	positions := []P{start}
	plant := grid[start.y][start.x]
	var p P
	for len(positions) > 0 {
		p, positions = positions[0], positions[1:]
		if pointsToRegions[p] > 0 {
			continue
		}
		pointsToRegions[p] = region
		for _, d := range DIRECTIONS {
			np := p.add(d)
			if np.y < 0 || np.y > ymax || np.x < 0 || np.x > xmax {
				continue
			}
			if grid[np.y][np.x] == plant {
				positions = append(positions, np)
			}
		}
	}
}

func sumPerimeterAndArea(regions []PointSet) int {
	sum := 0
	for _, points := range regions {
		perimeter := 0
		for p := range points {
			for _, d := range DIRECTIONS {
				if !points[p.add(d)] {
					perimeter++
				}
			}
		}
		sum += len(points) * perimeter
	}
	return sum
}

func sumSidesAndArea(regions []PointSet) int {
	sum := 0

	for _, points := range regions {
		corners := map[C]bool{}
		// Walk all edges until you reach every corner
		for p := range points {
			for _, d := range DIRECTIONS {
				if points[p.add(d)] { // not an edge
					continue
				}
				step := p
				dv := P{d.y, -d.x} // rotate counter-clockwise
				// walk until you reach a corner
				for points[step.add(dv)] && !points[step.add(d.add(dv))] {
					step = step.add(dv)
				}
				corners[C{step, d}] = true
			}
		}

		sum += len(corners) * len(points)
	}
	return sum
}

func main() {
	grid := parseInput()
	regions := findRegions(grid)

	println(sumPerimeterAndArea(regions))
	println(sumSidesAndArea(regions))
}
